### ISSUE TITLE
Define aims and outstanding technical tasks for distro support and Features-style config packages in core

### ISSUE THREAD

**Issue by nedjo** [backdrop/backdrop-issues#99]
Some interrelated issues with Drupal 7:
- On its own, a core Drupal install does practically nothing. Few if any sites run just the core.
- Rather than becoming more accessible, Drupal site development is increasingly costly. Typical site development costs range into the tens of thousands of dollars.
- Most custom development essentially repeats, with small variations, work that has been completed and paid for tens of thousands of times before.
- Alternatives to custom development like Drupal Gardens are limited, lacking a lot of the advantages of Drupal, and tied to particular corporate service offerings.
- With few exceptions, Drupal distributions seem to have low adoption rates.
- Despite early efforts like [Kit](https://drupal.org/project/kit) and later ones like [Debut](https://drupal.org/project/debut) and the interoperability aims in the [Open App Standard](https://groups.drupal.org/open-app-standard/oas) in practice, distro developers went their own ways, developing features that worked only within the confines of their own distro. A big missed potential of Drupal 7 was interoperable features.

The closest that Drupal 7 came to interoperability was the Panopoly feature set, which has been adopted by several distributions including Open Atrium 2. But, while it's a valuable contribution, Panopoly comes at significant costs. Far from being slim and basic, the base install of Panopoly adds dozens of modules to Drupal core and pushes it well past what's available in typical shared hosting.

It would be great to address these problems from the start in Backdrop. Steps to do so could include:
- Define clear standards for interoperable feature development.
- Ship core with several workable features that:
  - provide out of the box usable website solutions,
  - can be the basis for other, more specialized features and distributions,
  - model the standards for interoperable features.

**Comment by nedjo** [comment]
I posted a relevant Drupal 7 project this week that's similar to what I was suggesting above: https://www.drupal.org/sandbox/nedjo/2429233.
The basic approach:

* Offer a curated list of distributions and offers them for installation at install time.
* Download the one that's selected and extract to the profiles directory, then hand off the rest of the install process to that distro.

I'm interested in doing the same for Backdrop, and would be particularly interested if this could be considered for a future core release.

Also on the topic of distros....

Digging into preparing for distros in Drupal 8, I've run into some unexpected roadblocks that may affect Backdrop distros as well.

I've described the issues in a couple of [blog](http://chocolatelilyweb.ca/blog/drupal-8-configuration-management-what-about-small-sites-and-distributions) [posts](http://chocolatelilyweb.ca/blog/drupal-8-and-distributions-part-2-problems-and-prospects) and a [drupal.org issue](https://www.drupal.org/node/2414523).

In a nutshell: the CMI focus on the staging problem has created a number of barriers to distros, the main one arising from site configuration "living" on the site rather than in the providing modules.

I've begun work on a couple of relevant Drupal 8 modules: [Configuration Packager](https://www.drupal.org/project/config_packager) as a replacement for Features and [Configuration Synchronizer](https://www.drupal.org/sandbox/nedjo/2424835) for importing configuration updates from modules and themes. These might be relevant to Backdrop, though they would need a lot of adapting.

**Comment by quicksketch** [comment]
> the CMI focus on the staging problem has created a number of barriers to distros, the main one arising from site configuration "living" on the site rather than in the providing modules.

Right, the fact that config is stored in a central location instead of spread across modules makes module control of config more difficult. We understood that when taking it on, and expected that a contrib module would provide functionality to actually "bundle together" a collection of configuration to make a "feature". It would be nice if "Features" actually continued to live on, but did only what it was supposed to do in the first place, though I can understand wanting to avoid that particular name just because of the baggage.

One particular difference we have in our implementation (if I recall correctly) is that Backdrop still has a concept of "default" configuration. Views, Image Styles, and Layouts at least, all support "Reverting" configuration back into their default state. Unfortunately that a per-module implementation, not a universal system, but it could be extended to things like node types and (with a lot of work) fields.

> I'm interested in doing the same for Backdrop, and would be particularly interested if this could be considered for a future core release.

Streamlining the installation process has been an important goal for us. We actually eliminated the selection of both an install profile (by hiding the "minimal" install) and the database type (but that was a side effect of only supporting MySQL). In my experience, it's quite common for new users to choose what essentially amounts to the "wrong" profile or database right off the bat, and that can be a very negative experience for the user, causing them to eliminate Drupal (or Backdrop) as an option for their site.

However, with that said, I'm quite open to including the *functionality* to support distributions out of the box, as long we we don't present users with multiple (potentially "wrong") choices *by default*.

**Comment by quicksketch** [comment]
Pulling a quote from your blog post:

> In Drupal 8, module-provided configuration is imported once and once only--when the module is installed. The assumption is that, from that point onward, the configuration is "owned" by the site. 

Backdrop includes a "module" key for several types of config files. In the event that a module provided the config in the first place, some modules allow you to revert those back to their default. 

```php
/**
 * Revert the changes made by users to a default image style.
 *
 * @param style
 *   An image style array.
 * @return
 *   Boolean TRUE if the operation succeeded.
 */
function image_default_style_revert($style) {
  image_style_flush($style);
  config('image.style.' . $style['name'])->delete();
  config_install_default_config($style['module'], 'image.style.' . $style['name']);
  return TRUE;
}
```

The way these modules keep track of whether a module is in the "default" state is by setting a flag in the config file. The "default" configuration provided by the module has this config value set to a constant indicating it's in a "default" state. Once the user makes any changes, that flag is set to be in an "overridden" state. Then UI elements allow that particular config file to be reverted, via a mechanism like the function above.

It's not super-clean and it's not even consistent at this point. One of the things I really wanted to do before release was to make Views, Layouts, and Image Styles all use the same keys for keeping track of whether a config was in its default or overridden state, but shipping was more important. I'd love to see a consistent mechanism for handing default/overridden/custom configuration.

**Comment by nedjo** [comment]
> I'm quite open to including the functionality to support distributions out of the box, as long we we don't present users with multiple (potentially "wrong") choices by default.

Agreed with the aim here. The way I've written [Distribution Installer](https://www.drupal.org/sandbox/nedjo/2429233) is to require the profiles directory be writable at install time. (It might be better to reuse the FTP code used for in-browser module uploading, but I haven't done so yet. In D8 I'm not sure if that code's working, though I imagine it is in Backdrop.) One simple approach would be: a page to select a distro to install is presented only if the profiles directory is writable. So it becomes an option that's simple to invoke but not a default.

> In the event that a module provided the config in the first place, some modules allow you to revert those back to their default.

Nice. I considered using a flag to track the default status of config items. For [Configuration Synchronizer](https://www.drupal.org/sandbox/nedjo/2424835) I opted for using a snapshot for comparison.

There are a bunch of distinct issues here. Probably I should try to put together a Backdrop equivalent of the issue [META: Required functionality for Drupal 8 distributions](https://www.drupal.org/node/2414523). I'd need to look more closely at Backdrop code first.

**Comment by klonos** [comment]
Quite a few years later, I am wondering where we are with distribution support in https://github.com/backdrop-contrib.

Does one need to specify a .info file with `type = distribution` or something to have it be packaged? It certainly would not make any sense for an entire distro to be shown in the Project Browser, unless it was meant to be run as a "recipe", to download all required modules - but distros are more than a collection of modules. Nevertheless, we need a way to list them in b.org.
